### PR TITLE
Feature/public api create basic

### PR DESCRIPTION
**Pull request by noambergIL**
Please note : the waiter code needs refactor but we wanted the integration of the public-api (meaning block service runs the commit in txpool) loop to get in.

### CODE REVIEW

File: bootstrap/node_logic.go [hunk 1]
```diff
@@ -41,29 +41,28 @@ func NewNodeLogic(
...
 	processors[protocol.PROCESSOR_TYPE_NATIVE] = native.NewNativeProcessor(reporting)
 
 	crosschainConnectors := make(map[protocol.CrosschainConnectorType]services.CrosschainConnector)
 	crosschainConnectors[protocol.CROSSCHAIN_CONNECTOR_TYPE_ETHEREUM] = ethereum.NewEthereumCrosschainConnector()
 
-	virtualMachine := virtualmachine.NewVirtualMachine(blockStorage, stateStorage, processors, crosschainConnectors, reporting)
-	transactionPool := transactionpool.NewTransactionPool(ctx, gossip, virtualMachine, nodeConfig, reporting, primitives.TimestampNano(time.Now().UnixNano()))
-	publicApi := publicapi.NewPublicApi(transactionPool, virtualMachine, reporting)
-	consensusContext := consensuscontext.NewConsensusContext(transactionPool, virtualMachine, nil, nodeConfig, reporting)
+	gossipService := gossip.NewGossip(gossipTransport, nodeConfig, reporting)
```
[the remark this answers is not in the document]
Comment (noambergIL) [reply bootstrap/node_logic.go]: just before push my go land did a lint and told me that be fore gossip := gossip is a lint warning that i am using variable that might hide the package ... its only cosmetic and tal and i decided to change all names ... we didn't like g, bs, ss, vm, tp, as names ...

File: services/blockstorage/service.go [hunk 2]
```diff
@@ -359,6 +367,16 @@ func (s *service) syncBlockToStateStorage(committedBlockPair *protocol.BlockPair
 	return err
 }
 
+// TODO: this should not be called directly from CommitBlock, it should be called from a long living goroutine that continuously syncs the state storage
```
[the remark this answers is not in the document]
Comment (noambergIL) [reply services/blockstorage/service.go]: todo copied from function above ... this code should might be change together or not by Kiril/Jonathan

File: services/publicapi/waiter.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+	w.queue <- txWaiterMessage{
+		txId: txHash.KeyForMap(),
+		c:    receiptChannel,
+	}
+
+	return
+}
+
+func (w *txWaiter) tryEnqueue(message *txWaiterMessage) {
+	defer func() { recover() }()
```
[the remark this answers is not in the document]
Comment (noambergIL) [reply services/publicapi/waiter.go]: Mutex will avoid this. but:
avoiding edge cases of writing to closed queue ...

File: services/publicapi/service.go [hunk 4]
```diff
@@ -1,50 +1,111 @@
...
 
-	//TODO this is terrible, delete and make it right (shaiy)
+	if err != nil {
+		return prepareResponse(txResponse), errors.Errorf("error '%s' for transaction result", txResponse)
+	}
+	if txResponse.TransactionStatus == protocol.TRANSACTION_STATUS_DUPLCIATE_TRANSACTION_ALREADY_COMMITTED {
+		return prepareResponse(txResponse), nil
+	}
+
+	ta, err := waitContext.until(s.config.SendTransactionTimeout())
```
[the remark this answers is not in the document]
Comment (noambergIL) [reply services/publicapi/service.go]: going to be refactored so i'm ignoring for now,

File: services/publicapi/service.go [hunk 5]
```diff
@@ -1,50 +1,111 @@
...
+	if err != nil {
+		return prepareResponse(txResponse), err
+	}
+	return prepareResponse(ta), nil
+}
+
+func prepareResponse(transactionOutput *services.AddNewTransactionOutput) *services.SendTransactionOutput {
+	var receiptForClient *protocol.TransactionReceiptBuilder = nil
+
+	if receipt := transactionOutput.TransactionReceipt; receipt != nil {
```
[the remark this answers is not in the document]
Comment (noambergIL) [reply services/publicapi/service.go]: issue with builder ... opened to tal he knows.

**Comment by noambergIL** [comment]
Can you please merge it?

**Comment by electricmonk** [comment]
No, build consistently fails

Shai
________________________________

**Pull request merged**